### ISSUE TITLE
Fix Request: False positive for MDCPutWithoutRemove - put in a called method

### ISSUE THREAD

**Issue by hem-verma** [jborgers/PMD-jPinpoint-rules#473]
If we add a value to the MDC in the child method and remove it in the finally block of the parent method, It still complains about the issue. For this rule to work, value should be added and removed from MDC in the same method. Please check

public void A () {
try {
  MDC.put("a"); //Does not complain
  B();
} finally {
  MDC.remove("a");
  MDC.remove("b");
 }
}

private void B() {
  MDC.put("b"). //complain with Sonar issue
}

**Comment by stokpop** [comment]
Hi @hem-verma, thanks for this fix request. In the example given, the `B()` method is _public_ and there in no guarantee the `"b"` gets removed when not called via `A()` but by some external class and you might see MDC leakage to other threads when not removed properly by the external class. So rule is applicable. Does that make sense? Can your method `B()` be made private?

If `B()` is made private, then the class is responsible for adding and removing the MDC values (what might be called "Single Responsibility").

My suggestion for rule change: allow `MDC.put()` in _private_ methods in same class there is a `finally` with a `MDC.remove()`.

**Comment by hem-verma** [comment]
Hi @stokpop, Yes you are right. I will update the details. Actually in my case where its giving false positive, child method is private. Thanks for pointing out.

**Comment by jborgers** [comment]
I guess with `MDC.add` `MDC.put` was meant, right? I dont find `add` in the API.

**Comment by hem-verma** [comment]
Yes. Updated it. wrote it wrong at one place.

**Comment by jborgers** [comment]
And put has 2 arguments, like:
```
class Issue473 {
    public void foo() {
        try {
            MDC.put("a-key", "a-value"); // ok
            bar();
        } finally {
            MDC.remove("a-key");
            MDC.remove("b-key");
        }
    }

    private void bar() {
        MDC.put("b-key", "b-value"); //false positive
    }
}
```

**Comment by hem-verma** [comment]
Indeed. My bad. I tried to keep it as minimal as possible. :)

**Comment by jborgers** [comment]
Committed [fix](https://github.com/jborgers/PMD-jPinpoint-rules/commit/c5e6b64f5c8583073e619c353ab1c84e473e1c30)
